near-sdk-abi/src/private/mod.rs

```diff
@@ -0,0 +1,112 @@
+use std::path::{Path, PathBuf};
+
+use near_sdk::__private::AbiRoot;
+use quote::{format_ident, quote};
+use schemafy_lib::{Expander, Generator};
+
+pub fn generate_ext(
+    near_abi: AbiRoot,
+    contract_name: proc_macro2::Ident,
+) -> proc_macro2::TokenStream {
+    let schema_json = serde_json::to_string(&near_abi.abi.root_schema).unwrap();
+
+    let generator = Generator::builder().with_input_json(&schema_json).build();
+    let (mut token_stream, schema) = generator.generate_with_schema();
+    let mut expander = Expander::new(None, "", &schema);
+
+    let methods = near_abi
+        .abi
+        .functions
+        .iter()
+        .map(|m| {
+            let name = format_ident!("{}", m.name);
+            let result_type = m
+                .result
+                .clone()
+                .map(|r_param| {
+                    let r_type = expand_subschema(&mut expander, &r_param.type_schema);
+                    quote! { -> #r_type }
+                })
+                .unwrap_or_else(|| quote! {});
+            let args = m
+                .params
+                .iter()
+                .map(|a_param| {
+                    let a_type = expand_subschema(&mut expander, &a_param.type_schema);
+                    let a_name = format_ident!("{}", &a_param.name);
+                    quote! { #a_name: #a_type }
+                })
+                .collect::<Vec<_>>();
+            quote! { fn #name(&self, #(#args),*) #result_type; }
+        })
+        .collect::<Vec<_>>();
+
+    token_stream.extend(quote! {
+        #[near_sdk::ext_contract]
+        pub trait #contract_name {
+            #(#methods)*
+        }
+    });
+
+    token_stream
+}
+
+pub fn read_abi(abi_path: impl AsRef<Path>) -> AbiRoot {
+    let abi_path = if abi_path.as_ref().is_relative() {
+        let crate_root = get_crate_root().unwrap();
+        crate_root.join(&abi_path)
+    } else {
+        PathBuf::from(abi_path.as_ref())
+    };
+
+    let abi_json = std::fs::read_to_string(&abi_path)
```

Comment:
That was my first intuition as well, but unfortunately macro invocations are evaluated from outside to inside. So this means that `near_abi_ext!(include_str!("src/adder.json"))` evaluates `near_abi_ext!` first and it fails because `include_str!(...)` is not a string literal. Unless I am missing something here and you know of an alternative way to do this?